### PR TITLE
Adds ankerl::unordered_dense{segmented_map, segmented_set}

### PR DESCRIPTION
**Pull request by martinus**
This new underlying container has a much smoother memory allocation curve
than the default underlying `std::vector`.

* Much smoother memory usage, memory usage increases continuously.
* No high peak memory usage.
* Faster insertion because elements never need to be moved to new allocated blocks
* Slightly slower indexing compared to `std::vector` because an additional
  indirection is needed.

Abseil is fastest for this simple inserting test, taking a bit over 0.8 seconds.
It's peak memory usage is about 430 MB. Note how the memory usage goes down after
the last peak; when it goes down to ~290MB it has finished rehashing and could free
the previously used memory block.

`ankerl::unordered_dense::segmented_map` doesn't have these peaks, and instead has
a smooth increase of memory usage. Note there are still sudden drops & increases in
memory because the indexing data structure needs still needs to increase by a fixed
factor. But due to holding the data in a separate container we are able to first free
the old data structure, and then allocate a new, bigger indexing structure; thus we
do not have peaks.

bump to 4.0.0

### CODE REVIEW

File: include/ankerl/unordered_dense.h [hunk 1]
```diff
@@ -446,19 +442,315 @@ struct base_table_type_map {
...
+        return f;
+    }
+
+    using self_t = segmented_vector<T, Allocator, MaxSegmentSizeBytes>;
+    using vec_alloc = typename std::allocator_traits<Allocator>::template rebind_alloc<T*>;
+    static constexpr auto num_bits = num_bits_closest(MaxSegmentSizeBytes, sizeof(T));
+    static constexpr auto num_elements_in_block = 1U << num_bits;
+    static constexpr auto mask = num_elements_in_block - 1U;
+
+    std::vector<T*, vec_alloc> m_blocks{};
```
Comment (bigerl) [inline include/ankerl/unordered_dense.h:469]:
```suggestion
    vec_type<std::allocator_traits<T>::pointer, vec_alloc> m_blocks{};
```

`std::vector` does not work with allocators that require fancy pointers. 

Suggestion: make the type of m_blocks configurable via a template parameter

File: include/ankerl/unordered_dense.h [hunk 2]
```diff
@@ -446,19 +442,315 @@ struct base_table_type_map {
...
+    size_t m_size{};
+
+    /**
+     * Iterator class doubles as const_iterator and iterator
+     */
+    template <bool IsConst>
+    class iter_t {
+        using ptr_t = typename std::conditional_t<IsConst, T const* const*, T**>;
```
Comment (bigerl) [inline include/ankerl/unordered_dense.h:477]:
```suggestion
        using ptr_t = typename std::conditional_t<IsConst, std::allocator_traits<T>::pointer>::template rebind<std::allocator_traits<T>::pointer> const, std::pointer_traits<std::allocator_traits<T>::pointer>::template rebind<std::allocator_traits<T>::pointer>>;
```

File: include/ankerl/unordered_dense.h [hunk 3]
```diff
@@ -446,19 +442,315 @@ struct base_table_type_map {
...
+
+        template <bool B>
+        friend class iter_t;
+
+    public:
+        using difference_type = std::ptrdiff_t;
+        using value_type = T;
+        using reference = typename std::conditional<IsConst, value_type const&, value_type&>::type;
+        using pointer = typename std::conditional<IsConst, value_type const*, value_type*>::type;
+        using iterator_category = std::forward_iterator_tag;
```
Comment (bigerl) [inline include/ankerl/unordered_dense.h:489]: See line 587

**Review by bigerl — COMMENTED** [review]
For your consideration: Supporting Fancy Pointers

I've had a look at the implementation with regards to support allocators that use fancy pointers. This is especially relevant for `boost::interprocess::offset_ptr` which are used to work in shared memory and especially memory mapped files. 

I think the changes required to support allocators that use fancy pointers are minimal. I changed the relevant typedefs and function calls where I found them. The changes are untested and probably incomplete. 

A simple way to test that it works with offset pointers is either a thin wrapper around `std::allocator` that uses boost's offset pointers or to use [metall](https://github.com/LLNL/metall).

**Comment by martinus** [comment]
Hi @bigerl! Is the segmented vector of interest to you? It has a much more smooth allocation pattern than std:: vector.

I'll add more tests to make sure it works with fancy pointers 👍

**Comment by bigerl** [comment]
Hi @martinus. The map would be very interesting. A reference-stable hash map without the overhead of a node-based design sounds very promising. It makes it easy to store things there and still keep stable (offset-)pointers on them. That's something I was looking for in some of my projects for quite some time. 

In combination with good iteration speed (I would expect it to be similar to `tsl::sparse_map` with the segmented vector), fast lookups, amoderate memory footprint and fast insertions it should have very exposed spot along the Pareto front. ;)

**Comment by martinus** [comment]
Note that this implementation still won't have stable references, at least not on `erase()`.

**Comment by bigerl** [comment]
Ah, right. Just had a look into the code again. Tombstones would probably be the only option to maintain reference validity on erase. It comes with serious issues though: E.g., inserting entries 1...n and erasing entries 1...(n-1) would leaf the map with (n-1) tombstones. Most cases could probably be handled via intelligent bookkeeping involving merging of segments with values in distinct positions and removal of empty segments. 
It would make the code of the segment vector complex, probably slower and would not be an drop-in replacement for `std::vector` anymore. I guess it's out of discussion then. 

Having said that, also without stable references it would use it in shared memory. Especially the fact that the memory footprint stays always close to linear to the number of entries is very nice.

**Comment by martinus** [comment]
I could do stable references without tombstones, with an in-place freelist in the `segmented_vector`. This would give stable references without any lookup/insert performance penalty, erase would actually become much faster, only iterating would become slower. That's also the problem, iterating would need a complete rewrite and that's not really compatible with how it's currently done.

**Pull request merged**